src/goto-instrument/contracts/dynamic-frames/dfcc_instrument_loop.cpp

```diff
@@ -0,0 +1,717 @@
+    check_location.set_comment(
+      "Check invariant before entry for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    code_assertt assertion{initial_invariant};
+    assertion.add_source_location() = check_location;
+    converter.goto_convert(assertion, step_instrs, language_mode);
+  }
+
+  {
+    // check assigns clause inclusion with parent write set
+    // skip the check when if w_parent is NULL
+    auto goto_instruction = step_instrs.add(goto_programt::make_incomplete_goto(
+      equal_exprt(
+        outer_write_set,
+        null_pointer_exprt(to_pointer_type(outer_write_set.type()))),
+      loop_head_location));
+
+    auto check_var =
+      get_fresh_aux_symbol(
+        bool_typet(),
+        id2string(function_id),
+        "__check_assigns_clause_incl_loop_" + std::to_string(cbmc_loop_id),
+        loop_head_location,
+        language_mode,
+        symbol_table)
+        .symbol_expr();
+
+    step_instrs.add(goto_programt::make_decl(check_var, loop_head_location));
+    step_instrs.add(goto_programt::make_function_call(
+      library.write_set_check_assigns_clause_inclusion_call(
+        check_var, outer_write_set, addr_of_loop_write_set, loop_head_location),
+      loop_head_location));
+
+    source_locationt check_location(loop_head_location);
+    check_location.set_property_class("loop_assigns");
+    check_location.set_comment(
+      "Check assigns clause inclusion for loop " +
+      id2string(check_location.get_function()) + "." +
+      std::to_string(cbmc_loop_id));
+    step_instrs.add(goto_programt::make_assertion(check_var, check_location));
+    step_instrs.add(goto_programt::make_dead(check_var, loop_head_location));
+
+    auto label_instruction =
+      step_instrs.add(goto_programt::make_skip(loop_head_location));
+    goto_instruction->complete_goto(label_instruction);
+  }
+
+  {
+    // Generate havocing code for assigns targets.
+    const auto in_loop_havoc_block =
+      new_tmp_symbol(
+        bool_typet(),
+        loop_head_location,
+        language_mode,
+        symbol_table,
+        std::string(IN_LOOP_HAVOC_BLOCK) + +"__" + std::to_string(cbmc_loop_id))
+        .symbol_expr();
+    step_instrs.add(
+      goto_programt::make_decl(in_loop_havoc_block, loop_head_location));
+    step_instrs.add(
+      goto_programt::make_assignment(in_loop_havoc_block, true_exprt{}));
+    step_instrs.destructive_append(havoc_instrs);
+    step_instrs.add(
+      goto_programt::make_assignment(in_loop_havoc_block, false_exprt{}));
+  }
+
+  {
+    // Assume the loop invariant after havocing the state
+    code_assumet assumption{invariant};
+    assumption.add_source_location() = loop_head_location;
+    converter.goto_convert(assumption, step_instrs, language_mode);
+  }
+
+  {
+    // Generate assignments to store the multidimensional decreases clause's
+    // value just before the loop_head
+    if(!decreases_clauses.empty())
+    {
```

Comment:
I think this is unnecessary (and then you won't need the `decreases_clauses` parameter).